Lib/compileall.py

```diff
@@ -227,12 +227,24 @@ def compile_file(fullname, ddir=None, force=False, rx=None, quiet=0,
         if tail == '.py':
             if not force:
                 try:
-                    mtime = int(os.stat(fullname).st_mtime)
-                    expect = struct.pack('<4sLL', importlib.util.MAGIC_NUMBER,
-                                         0, mtime & 0xFFFF_FFFF)
                     for cfile in opt_cfiles.values():
                         with open(cfile, 'rb') as chandle:
-                            actual = chandle.read(12)
+                            header = chandle.read(16)
+
+                        if header[4]:
```

Comment:
You can check the magic number without reading the source file.

Also, header[4] can raise IndexError. And you need to check other bytes and bits.

```suggestion
                        if len(header) < 16 or header[:4] != importlib.util.MAGIC_NUMBER or struct.unpack('<L', header[4:8])[0] & ~0b11:
                            break
                        if header[4] & 0b1:
```